wgengine/magicsock/magicsock_linux.go

```diff
@@ -38,8 +46,24 @@ var debugRawDiscoReads = envknob.RegisterBool("TS_DEBUG_RAW_DISCO")
 // These are our BPF filters that we use for testing packets.
 var (
 	magicsockFilterV4 = []bpf.Instruction{
-		// For raw UDPv4 sockets, BPF receives the entire IP packet to
-		// inspect.
+		// For raw sockets (with ETH_P_IP set), the BPF program
+		// receives the entire IPv4 packet, but not the Ethernet
+		// header.
+
+		// First off, verify that we have a packet that's big enough to
+		// (possibly) contain a disco packet.
+		//
+		// The length of an IPv4 disco packet is composed of:
+		// - 20 bytes for the IPv4 header
```

Comment:
There's a small catch here: the code prior to this change supports IPv4 packets with extensions, whereas this check assumes there are none. The `bpf.LoadMemShift` below is what does that magic, by extracting the "number of 4-byte words in the header" field and multiplying it out to get the offset of the next header.

I'd suggest doing the payload check after the LoadMemShift, with a `bpf.LoadIndirect` and the length of just UDP + disco headers. That'll do a load relative to the end of the IPv4 header, regardless of its length.

The cBPF VM is an accumulator machine with A and X registers (main accumulator and indirection offset register respectively), and LoadMemShift implicitly loads into the X register. So, I think you can insert a second LoadIndirect between the current LoadMemShift and the magic number check, and that'll all work out. i.e.:

```go
// X = index_of_first_byte_after_IP_header
bpf.LoadMemShift{Off:0}
// A = last_payload_byte_of_minimal_packet
bpf.LoadIndirect{Off: udpHeaderSize + discoMinHeaderSize - 1, Size: 1}
// A = first_half_of_disco_magic
bpf.LoadIndirect{Off: udpHeaderSize, Size: 4}
...
```

For peak paranoia, we'd want to check that the cBPF->eBPF translator and the eBPF JIT don't optimize away that middle LoadIndirect... But since loads in cBPF have implicit range checks, doing so would alter program semantics in an incorrect way, so I'd expect the cBPF translator to materialize the right explicit eBPF test and branch, and be unable to JIT it away?...

Reply:
Done, and good call; definitely worth being consistent with our length checks